File: app/src/main/java/workshop1024/com/xproject/model/filter/source/FilterDataSource.java
package workshop1024.com.xproject.model.filter.source;

import java.util.List;

import workshop1024.com.xproject.model.filter.Filter;

public interface FilterDataSource {
    void getFilters(LoadFiltersCallback loadFiltersCallback);

    void addFilterByName(String filterName);

    void deleteFilterById(String filterId);

    interface LoadFiltersCallback {
        void onPublishersLoaded(List<Filter> filterList);

        void onDataNotAvailable();
    }
}

File: main/src/main/java/workshop1024/com/xproject/main/model/filter/source/FilterDataSource.kt
package workshop1024.com.xproject.main.model.filter.source

import workshop1024.com.xproject.main.model.filter.Filter

interface FilterDataSource {
    fun getFilters(loadCallback: LoadCallback)

    fun addFilter(filter: Filter)

    fun deleteFilterById(filterId: String)

    fun deleteAllFilters()

    fun getIsRequestRemote(): Boolean

    interface LoadFiltersCallback : LoadRemoteFiltersCallback, LoadCacheOrLocalFiltersCallback

    interface LoadCacheOrLocalFiltersCallback : LoadCallback {
        fun onCacheOrLocalFiltersLoaded(filterList: List<Filter>)
    }

    interface LoadRemoteFiltersCallback : LoadCallback {
        fun onRemoteFiltersLoaded(filterList: List<Filter>)
    }

    interface LoadCallback {
        fun onDataNotAvailable()
    }
}
